### PR TITLE
[fix](session-variable) fix bug that checkpoint may overwrite the global variables

### PR DESCRIPTION
**Pull request by morningman**
## Proposed changes

Close related #7525 (replace it with issue number if it exists).

We should create temporary object for some static fields when doing checkpoint,
to avoid these variables to be overwritten by the checkpoint process.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java [hunk 1]
```diff
@@ -304,17 +286,30 @@ public static void setLowerCaseTableNames(int mode) throws DdlException {
 
     // global variable persistence
     public static void write(DataOutputStream out) throws IOException {
-        defaultSessionVariable.write(out);
+        SessionVariable variablesToWrite = defaultSessionVariable;
+        if (Catalog.isCheckpointThread()) {
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java:290]: There should not be non-checkpoint-thread calling this method. Why not directly use defaultSessionVariableForCkpt instead.

File: fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java [hunk 2]
```diff
@@ -304,17 +286,30 @@ public static void setLowerCaseTableNames(int mode) throws DdlException {
...
+        GlobalVarPersistInfo info = new GlobalVarPersistInfo(variablesToWrite, varNames);
         info.write(out);
     }
 
     public static void read(DataInputStream in) throws IOException, DdlException {
         wlock.lock();
         try {
-            defaultSessionVariable.readFields(in);
+            SessionVariable variablesToRead = defaultSessionVariable;
+            if (Catalog.isCheckpointThread()) {
```
Comment (xy720) [inline fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java:306]: The same as above

File: fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java (hunk 1 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java:290]: Not exactly, the `write` method may be called in:
1. Checkpoint thread
2. when calling `/dump` http api

File: fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java (hunk 2 above)
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/qe/VariableMgr.java:306]: No, the `read` method will be called:
1. when FE start
2. checkpoint thread

**Review by xy720 — APPROVED** [review]
LGTM

**Pull request merged**